src/payment/models/pricing/utils/price-step.ts

```diff
@@ -0,0 +1,165 @@
+  }
+
+  //*** HELPER METHODS ***//
+
+  private getFixedPrice(fromCurrency: string, toCurrency: string): [Price, PriceProviderName] {
+    const price = new Price();
+
+    price.source = fromCurrency;
+    price.target = toCurrency;
+    price.price = this.options.fixedPrice;
+
+    return [price, 'FixedPrice'];
+  }
+
+  private async getMatchingPrice(
+    fromCurrency: string,
+    toCurrency: string,
+    matchThreshold = 0.005,
+  ): Promise<[Price, PriceProviderName]> {
+    const [primaryPrice, primaryProvider] = await this.getPrimaryPrice(fromCurrency, toCurrency);
+
+    try {
+      const [refPrice, refSource] = await this.getReferencePrice(fromCurrency, toCurrency);
+
+      const { price: _mainPrice } = primaryPrice;
+      const { price: _refPrice } = refPrice;
+
+      if (Math.abs(_refPrice - _mainPrice) / _mainPrice > matchThreshold)
+        throw new PriceMismatchException(
+          `${fromCurrency} to ${toCurrency} price mismatch (${primaryProvider}: ${_mainPrice}, ${refSource}: ${_refPrice})`,
+        );
+    } catch (e) {
+      if (e instanceof PriceMismatchException) throw e;
+
+      console.warn(
+        `Proceeding without reference check at: ${this.options.providers.reference.map(
+          (p) => p.name,
+        )}. From ${fromCurrency} to ${toCurrency}`,
+      );
+    }
+
+    return [primaryPrice, primaryProvider];
+  }
+
+  private async getPrimaryPrice(fromCurrency: string, toCurrency: string): Promise<[Price, PriceProviderName]> {
+    const primaryProviders = this.options.providers.primary;
+
+    const [price, providerName] = await this.tryProviders(fromCurrency, toCurrency, primaryProviders);
+
+    if (!price) {
+      throw new Error(
+        `Could not find primary price at: ${primaryProviders.map(
+          (p) => p.name + ' ',
+        )}. From ${fromCurrency} to ${toCurrency}`,
+      );
+    }
+
+    return [price, providerName];
+  }
+
+  private async getReferencePrice(fromCurrency: string, toCurrency: string): Promise<[Price, PriceProviderName]> {
+    const referenceProviders = this.options.providers.reference;
+
+    const [price, providerName] = this.options.referenceTo
+      ? await this.tryProviders(fromCurrency, this.options.referenceTo, referenceProviders)
+      : await this.tryProviders(fromCurrency, toCurrency, referenceProviders);
+
+    if (!price) {
+      throw new Error(
+        `Could not find reference price at: ${referenceProviders.map(
+          (p) => p.name + ';',
+        )}. From ${fromCurrency} to ${toCurrency}`,
+      );
+    }
+
+    return [price, providerName];
+  }
+
+  private async tryProviders(
+    fromCurrency: string,
+    toCurrency: string,
+    providers: PriceProvider[] = [],
+  ): Promise<[Price, PriceProviderName]> {
+    for (const provider of providers) {
+      try {
+        return [await provider.getPrice(fromCurrency, toCurrency), provider.name];
+      } catch {
+        return null;
```

Comment:
Is this correct? Don't we leave the for loop?

Reply:
no, it's a bug indeed. Thanks again for spotting! Fixed